src/mito2/src/read/scan_region.rs

```diff
@@ -315,3 +356,235 @@ fn file_in_range(file: &FileHandle, predicate: &TimestampRange) -> bool {
+        for mem in &self.memtables {
+            let iter = mem.iter(Some(self.mapper.column_ids()), self.predicate.clone())?;
+            sources.push(Source::Iter(iter));
+        }
+        for file in &self.files {
+            let maybe_reader = self
+                .access_layer
+                .read_sst(file.clone())
+                .predicate(self.predicate.clone())
+                .time_range(self.time_range)
+                .projection(Some(self.mapper.column_ids().to_vec()))
+                .cache(self.cache_manager.clone())
+                .index_applier(self.index_applier.clone())
+                .build()
+                .await;
+            let reader = match maybe_reader {
+                Ok(reader) => reader,
+                Err(e) => {
+                    if e.is_object_not_found() && self.ignore_file_not_found {
+                        error!(e; "File to scan does not exist, region_id: {}, file: {}", file.region_id(), file.file_id());
+                        continue;
+                    } else {
+                        return Err(e);
+                    }
+                }
+            };
+            if compat::has_same_columns(self.mapper.metadata(), reader.metadata()) {
+                sources.push(Source::Reader(Box::new(reader)));
+            } else {
+                // They have different schema. We need to adapt the batch first so the
+                // mapper can convert it.
+                let compat_reader =
+                    CompatReader::new(&self.mapper, reader.metadata().clone(), reader)?;
+                sources.push(Source::Reader(Box::new(compat_reader)));
+            }
+        }
+
+        READ_SST_COUNT.observe(self.files.len() as f64);
+
+        Ok(sources)
+    }
+
+    /// Scans sources in parallel.
+    ///
+    /// # Panics if the input doesn't allow parallel scan.
+    pub(crate) async fn build_parallel_sources(&self) -> Result<Vec<Source>> {
+        assert!(self.parallelism.allow_parallel_scan());
+        // Scall all memtables and SSTs.
+        let sources = self.build_sources().await?;
+        let semaphore = Arc::new(Semaphore::new(self.parallelism.parallelism));
+        // Spawn a task for each source.
+        let sources = sources
+            .into_iter()
+            .map(|source| {
+                let (sender, receiver) = mpsc::channel(self.parallelism.channel_size);
+                self.spawn_scan_task(source, semaphore.clone(), sender);
+                let stream = Box::pin(ReceiverStream::new(receiver));
+                Source::Stream(stream)
+            })
+            .collect();
+        Ok(sources)
+    }
+
+    /// Scans the input source in another task and sends batches to the sender.
+    pub(crate) fn spawn_scan_task(
+        &self,
+        mut input: Source,
+        semaphore: Arc<Semaphore>,
+        sender: mpsc::Sender<Result<Batch>>,
+    ) {
+        tokio::spawn(async move {
```

Comment:
Maybe use dedicated `common_read` runtime?
```suggestion
        common_runtime::spawn_read(async move {
```

Reply:
I fixed all comments in this commit https://github.com/GreptimeTeam/greptimedb/pull/3598/commits/7760a967a493fed1fe0f7bbd02f61da39f577ab1